### ISSUE TITLE
Vector as an action

### ISSUE THREAD

**Issue by tms1337** [reinforceio/tensorforce#152]
Hello,

I implemented my own environment and I want each action to be a vector of n+1 values in range [0, 1]. My states and actions properties look like this. Note that I am only beginning so I might be doing something terribly wrong.

   ```
@property
    def states(self):
        return dict(shape=(self.n, self.window, self.feature_n),
                    type='float')

    @property
    def actions(self):
        return { str(i): dict(continuous=True, min_value=0, max_value=1) for i in range(self.n + 1) }
```
I want my action space to be constrained to sum to 1. I figured I would use softmax as the last layer of my network, so in my agent definition I have this (I use VPGAgent):

```
network=layered_network_builder([
            dict(type='conv2d', size=5, window=3),
            dict(type='flatten'),
            dict(type='dense', size=n+1, activation='softmax')
        ])
```
The problem is I get n+1 actions that are in range [0, 1] but softmax seems to be ignored. Does this mean the network is applied for each action separately, so softmax in the last layer has no effect? 

 **If so what is the prefered way of having a vector action**. I tried 

`dict(continuous=True, min_value=0, max_value=1, num_actions=n+1)`

but this seems not to work. I did not really expect it to work, but I am aiming at something similar to this.

I managed to softmax it in execute() method of an environment, but I consider this hacky and not the right way to do this.

```
        ...
        action = sorted(action.items(), key=lambda x: x[0])
        action = np.array([ k[1] for k in action ])

        action = np.exp(action)
        action /= sum(action)
```

**Comment by AlexKuhnle** [comment]
Hi @tms1337,

The first code snippet looks good and, as I understand it, indeed gives you a vector action between 0 and 1. Note that the parameter `num_actions` is only valid for discrete actions, where it determines the number of discrete alternatives. Further note that the output of the network does not directly correspond to the output of the model, but is used as *'conditioning'* input to the policy distribution, which then emits the action based on it.

The *'problem'* is that we currently do not have constraints between actions within a vector, so each value will be independently sampled between 0 and 1. To achieve what you want the best way, currently, is indeed to do a softmax *'as part of the environment'*. In this case you probably don't even want to constrain the actions to be in [0, 1]. Personally, I don't think this way is too bad / hacky, actions as *'logits'* instead of *'probabilities'*.

I'm very sceptical we ever will have arbitrary constraints on action as part of the action definition, but rather leave it to the environment. However, we are currently working towards releasing a new version, which will change and improve a lot of things, one being that actions are not just discrete and continuous anymore, but have a *'type'*. It's definitely conceivable that one such type could be 'distribution' or so, which internally uses Gaussians but applies a softmax before outputting the vector. Worth trying once the new update is out.

**Comment by tms1337** [comment]
@AlexKuhnle Thanks for the quick answer. I use softmax since my action represents the distribution of assets of some sort. I know this might be in the domain of RL theory, but would it be decremental for the performance or convergence of the agent that there are several possible actions that correspond to the same effective action? What I mean is that several possible actions that agent can generate can result in the same asset distribution. What are your thoughts on this?

**Comment by AlexKuhnle** [comment]
My intuition -- but I'm definitely not an expert on this -- is that it won't matter too much. What the agent would need to learn is the *relative* effect of choosing a logit action, in the context of the other logit action choices, since the absolute value is not as meaningful, as you say. I can see that there might be better ways, but nothing immediately comes to my mind, and I would expect the discussed method to handle this reasonably well.
